<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>lec_flow_push – integration_n_clustering</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script>
<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


<link rel="stylesheet" href="../styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">lec_flow_push</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../index.html" rel="" target="">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../about.html" rel="" target="">
 <span class="menu-text">About</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-preprocessing" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">preprocessing</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-preprocessing">    
        <li>
    <a class="dropdown-item" href="../code/convert_to_adata.html" rel="" target="">
 <span class="dropdown-text">convert_to_adata.html</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../code/preprocessing_n_removal.html" rel="" target="">
 <span class="dropdown-text">preprocessing_n_removal.html</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../code/integration_n_clustering.html" rel="" target="">
 <span class="dropdown-text">integration_n_clustering.html</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../code/clusters_markers.html" rel="" target="">
 <span class="dropdown-text">clusters_markers.html</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-differential-expression" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Differential expression</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-differential-expression">    
        <li>
    <a class="dropdown-item" href="../code/vis_de.html" rel="" target="">
 <span class="dropdown-text">vis_de</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../code/pb_de.html" rel="" target="">
 <span class="dropdown-text">pb_de</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-interactively" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Interactively</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-interactively">    
        <li>
    <a class="dropdown-item" href="../code/shinycell.html" rel="" target="">
 <span class="dropdown-text">shinycell</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-exploratory-data-analysis" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Exploratory data analysis</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-exploratory-data-analysis">    
        <li>
    <a class="dropdown-item" href="../code/data_viz.html" rel="" target="">
 <span class="dropdown-text">data_viz</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../code/valve_sub.html" rel="" target="">
 <span class="dropdown-text">valve subgroups</span></a>
  </li>  
    </ul>
  </li>
</ul>
            <div class="quarto-navbar-tools ms-auto">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">



<div class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scanpy <span class="im">as</span> sc</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> matplotlib <span class="im">import</span> pyplot <span class="im">as</span> plt</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>sc.settings.set_figure_params(</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>    dpi<span class="op">=</span><span class="dv">80</span>,</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>    facecolor<span class="op">=</span><span class="st">"white"</span>,</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>    frameon<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="2">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> sc.read_h5ad(<span class="st">"../output/0.2-sce-after-contamination-removal.h5ad"</span>) <span class="co"># read adata</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>adata.obs <span class="op">=</span> adata.obs[[<span class="st">'Sample'</span>, <span class="st">'Barcode'</span>, <span class="st">'donor'</span>, <span class="st">'tissue'</span>, <span class="st">'tissue_origin'</span>, <span class="st">'sequencing_run'</span>, <span class="st">'sex'</span>]].astype(<span class="bu">str</span>) <span class="co"># take just the useful info</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>sc.pp.calculate_qc_metrics(adata, inplace<span class="op">=</span><span class="va">True</span>) <span class="co"># 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', etc...</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>adata.obs.loc[adata.obs.donor<span class="op">==</span><span class="st">"5.0"</span>, <span class="st">"sex"</span>] <span class="op">=</span> <span class="st">"M"</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>adata.obs.loc[adata.obs.donor<span class="op">==</span><span class="st">"7.0"</span>, <span class="st">"sex"</span>] <span class="op">=</span> <span class="st">"F"</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>adata</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="2">
<pre><code>AnnData object with n_obs × n_vars = 21588 × 18627
    obs: 'Sample', 'Barcode', 'donor', 'tissue', 'tissue_origin', 'sequencing_run', 'sex', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes'
    var: 'ID', 'Symbol', 'Type', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts'
    uns: 'X_name'
    layers: 'logcounts'</code></pre>
</div>
</div>
<div class="cell" data-execution_count="3">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># cells stratification</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>adata.obs.groupby([<span class="st">"sequencing_run"</span>, <span class="st">"donor"</span>, <span class="st">"tissue_origin"</span>, <span class="st">"tissue"</span>])[[<span class="st">"Barcode"</span>]].count().replace(<span class="dv">0</span>, np.nan).dropna().astype(<span class="bu">int</span>).to_csv(<span class="st">"../output/1.0-cells_after_filtering.csv"</span>)</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>adata.obs.groupby([<span class="st">"sequencing_run"</span>, <span class="st">"donor"</span>, <span class="st">"tissue_origin"</span>, <span class="st">"tissue"</span>])[[<span class="st">"Barcode"</span>]].count().replace(<span class="dv">0</span>, np.nan).dropna().astype(<span class="bu">int</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="3">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">Barcode</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">sequencing_run</th>
<th data-quarto-table-cell-role="th">donor</th>
<th data-quarto-table-cell-role="th">tissue_origin</th>
<th data-quarto-table-cell-role="th">tissue</th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td rowspan="3" data-quarto-table-cell-role="th" data-valign="top">1.0</td>
<td rowspan="2" data-quarto-table-cell-role="th" data-valign="top">1.0</td>
<td rowspan="2" data-quarto-table-cell-role="th" data-valign="top">abdomen</td>
<td data-quarto-table-cell-role="th">fat</td>
<td>158</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">skin</td>
<td>3784</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2.0</td>
<td data-quarto-table-cell-role="th">arm</td>
<td data-quarto-table-cell-role="th">fat</td>
<td>2794</td>
</tr>
<tr class="even">
<td rowspan="4" data-quarto-table-cell-role="th" data-valign="top">2.0</td>
<td rowspan="2" data-quarto-table-cell-role="th" data-valign="top">3.0</td>
<td rowspan="2" data-quarto-table-cell-role="th" data-valign="top">abdomen</td>
<td data-quarto-table-cell-role="th">fat</td>
<td>724</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">skin</td>
<td>540</td>
</tr>
<tr class="even">
<td rowspan="2" data-quarto-table-cell-role="th" data-valign="top">4.0</td>
<td rowspan="2" data-quarto-table-cell-role="th" data-valign="top">abdomen</td>
<td data-quarto-table-cell-role="th">fat</td>
<td>49</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">skin</td>
<td>584</td>
</tr>
<tr class="even">
<td rowspan="4" data-quarto-table-cell-role="th" data-valign="top">3.0</td>
<td rowspan="2" data-quarto-table-cell-role="th" data-valign="top">6.0</td>
<td rowspan="2" data-quarto-table-cell-role="th" data-valign="top">abdomen</td>
<td data-quarto-table-cell-role="th">fat</td>
<td>1974</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">skin</td>
<td>1692</td>
</tr>
<tr class="even">
<td rowspan="2" data-quarto-table-cell-role="th" data-valign="top">7.0</td>
<td rowspan="2" data-quarto-table-cell-role="th" data-valign="top">thigh</td>
<td data-quarto-table-cell-role="th">fat</td>
<td>1680</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">skin</td>
<td>2847</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">4.0</td>
<td data-quarto-table-cell-role="th">5.0</td>
<td data-quarto-table-cell-role="th">abdomen</td>
<td data-quarto-table-cell-role="th">mixed</td>
<td>4762</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<div class="cell" data-execution_count="4">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># standardize logcounts</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>adata.layers[<span class="st">"logcounts.scaled"</span>] <span class="op">=</span> sc.pp.scale(adata, zero_center<span class="op">=</span><span class="va">True</span>, max_value<span class="op">=</span><span class="va">None</span>, copy<span class="op">=</span><span class="va">True</span>, layer<span class="op">=</span><span class="st">"logcounts"</span>).layers[<span class="st">"logcounts"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="5">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># select the HVGs using pearson method</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>pearson_hvgs <span class="op">=</span> sc.experimental.pp.highly_variable_genes(adata, inplace<span class="op">=</span><span class="va">False</span>, batch_key<span class="op">=</span><span class="st">"donor"</span>, layer<span class="op">=</span><span class="va">None</span>, flavor<span class="op">=</span><span class="st">"pearson_residuals"</span>, n_top_genes<span class="op">=</span><span class="dv">500</span>)</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>pearson_hvgs</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="5">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">means</th>
<th data-quarto-table-cell-role="th">variances</th>
<th data-quarto-table-cell-role="th">residual_variances</th>
<th data-quarto-table-cell-role="th">highly_variable_rank</th>
<th data-quarto-table-cell-role="th">highly_variable_nbatches</th>
<th data-quarto-table-cell-role="th">highly_variable_intersection</th>
<th data-quarto-table-cell-role="th">highly_variable</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">ENSG00000187634.SAMD11</td>
<td>0.000973</td>
<td>0.000972</td>
<td>0.908626</td>
<td>NaN</td>
<td>0</td>
<td>False</td>
<td>False</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">ENSG00000188976.NOC2L</td>
<td>0.042709</td>
<td>0.049040</td>
<td>1.066757</td>
<td>NaN</td>
<td>0</td>
<td>False</td>
<td>False</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">ENSG00000187961.KLHL17</td>
<td>0.005420</td>
<td>0.006224</td>
<td>1.313732</td>
<td>NaN</td>
<td>0</td>
<td>False</td>
<td>False</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">ENSG00000187583.PLEKHN1</td>
<td>0.001019</td>
<td>0.001018</td>
<td>0.832943</td>
<td>NaN</td>
<td>0</td>
<td>False</td>
<td>False</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">ENSG00000187642.PERM1</td>
<td>0.000232</td>
<td>0.000510</td>
<td>0.522404</td>
<td>NaN</td>
<td>0</td>
<td>False</td>
<td>False</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">ENSG00000271254.ENSG00000271254</td>
<td>0.011349</td>
<td>0.013074</td>
<td>1.100826</td>
<td>NaN</td>
<td>0</td>
<td>False</td>
<td>False</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">ENSG00000276345.ENSG00000276345</td>
<td>0.025246</td>
<td>0.029983</td>
<td>0.637354</td>
<td>NaN</td>
<td>0</td>
<td>False</td>
<td>False</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">ENSG00000276017.ENSG00000276017</td>
<td>0.000046</td>
<td>0.000046</td>
<td>0.083079</td>
<td>NaN</td>
<td>0</td>
<td>False</td>
<td>False</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">ENSG00000278817.ENSG00000278817</td>
<td>0.008894</td>
<td>0.009093</td>
<td>1.044582</td>
<td>NaN</td>
<td>0</td>
<td>False</td>
<td>False</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">ENSG00000277196.ENSG00000277196</td>
<td>0.000139</td>
<td>0.000139</td>
<td>0.427159</td>
<td>NaN</td>
<td>0</td>
<td>False</td>
<td>False</td>
</tr>
</tbody>
</table>

<p>18627 rows × 7 columns</p>
</div>
</div>
</div>
<div class="cell" data-execution_count="6">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>plt.figure(figsize<span class="op">=</span>(<span class="dv">10</span>,<span class="dv">10</span>))</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>plt.colorbar(plt.scatter(</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>    pearson_hvgs.means, pearson_hvgs.variances,</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>    alpha<span class="op">=</span><span class="fl">.5</span>, s<span class="op">=</span>pearson_hvgs.residual_variances<span class="op">*</span><span class="dv">10</span>, c<span class="op">=</span>np.log1p(pearson_hvgs.residual_variances),</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>    cmap<span class="op">=</span><span class="st">"spring_r"</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>), label<span class="op">=</span><span class="st">"log(residuals)"</span>)</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>plt.xscale(<span class="st">"log"</span>)</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>plt.yscale(<span class="st">"log"</span>)</span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">"mean"</span>)</span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">"variance"</span>)</span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="integration_n_clustering_files/figure-html/cell-7-output-1.png" class="img-fluid"></p>
</div>
</div>
<div class="cell" data-execution_count="7">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>plt.figure(figsize<span class="op">=</span>(<span class="dv">10</span>,<span class="dv">10</span>))</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>plt.colorbar(plt.scatter(</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>    pearson_hvgs.means, pearson_hvgs.variances,</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>    alpha<span class="op">=</span><span class="fl">.25</span>, s<span class="op">=</span>pearson_hvgs.residual_variances<span class="op">*</span><span class="dv">10</span>, c<span class="op">=</span>pearson_hvgs.highly_variable_nbatches,</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>    cmap<span class="op">=</span><span class="st">"turbo"</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>), label<span class="op">=</span><span class="st">"number of batches"</span>)</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>plt.xscale(<span class="st">"log"</span>)</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>plt.yscale(<span class="st">"log"</span>)</span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">"mean"</span>)</span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">"variance"</span>)</span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="integration_n_clustering_files/figure-html/cell-8-output-1.png" class="img-fluid"></p>
</div>
</div>
<div class="cell" data-execution_count="8">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>plt.figure(figsize<span class="op">=</span>(<span class="dv">10</span>,<span class="dv">5</span>))</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">"How many genes are highly valiable in at least n batches?"</span>)</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>plt.plot(</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>    np.arange(adata.obs.donor.nunique())<span class="op">+</span><span class="dv">1</span>,</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>    (pearson_hvgs[[<span class="st">"highly_variable_nbatches"</span>]].values <span class="op">&gt;</span> np.arange(adata.obs.donor.nunique())[<span class="va">None</span>,:]).<span class="bu">sum</span>(axis<span class="op">=</span><span class="dv">0</span>),</span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>    linewidth<span class="op">=</span><span class="dv">5</span></span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">"n genes"</span>)</span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">"n batches"</span>)</span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="integration_n_clustering_files/figure-html/cell-9-output-1.png" class="img-fluid"></p>
</div>
</div>
<div class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># reading the LECs population and subpopulation markers</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>markers <span class="op">=</span> pd.read_csv(<span class="st">"../markers.csv"</span>, header<span class="op">=</span><span class="va">None</span>)</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>markers.columns <span class="op">=</span> [<span class="st">"gene_ID"</span>, <span class="st">"population"</span>, <span class="st">"population_ID"</span>, <span class="st">"positive"</span>, <span class="st">"known"</span>]</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>markers <span class="op">=</span> markers.loc[markers.population.<span class="bu">str</span>.match(<span class="st">".*LEC.*"</span>)]</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>markers[<span class="st">"positive"</span>] <span class="op">=</span> markers.positive<span class="op">==</span><span class="st">"positive"</span></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>markers</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="9">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">gene_ID</th>
<th data-quarto-table-cell-role="th">population</th>
<th data-quarto-table-cell-role="th">population_ID</th>
<th data-quarto-table-cell-role="th">positive</th>
<th data-quarto-table-cell-role="th">known</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>ENSG00000117707.PROX1</td>
<td>LECs</td>
<td>LEC</td>
<td>True</td>
<td>yes</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>ENSG00000162493.PDPN</td>
<td>LECs</td>
<td>LEC</td>
<td>True</td>
<td>yes</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>ENSG00000133800.LYVE1</td>
<td>capillary-LEC</td>
<td>cpLEC</td>
<td>True</td>
<td>yes</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>ENSG00000137077.CCL21</td>
<td>capillary-LEC</td>
<td>cpLEC</td>
<td>True</td>
<td>yes</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>ENSG00000189056.RELN</td>
<td>capillary-LEC</td>
<td>cpLEC</td>
<td>True</td>
<td>yes</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">5</td>
<td>ENSG00000129048.ACKR4</td>
<td>capillary-LEC</td>
<td>cpLEC</td>
<td>False</td>
<td>yes</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">6</td>
<td>ENSG00000108691.CCL2</td>
<td>capillary-LEC</td>
<td>cpLEC</td>
<td>False</td>
<td>no</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">7</td>
<td>ENSG00000081041.CXCL2</td>
<td>capillary-LEC</td>
<td>cpLEC</td>
<td>False</td>
<td>no</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">8</td>
<td>ENSG00000133800.LYVE1</td>
<td>precollector-LEC</td>
<td>pcLEC</td>
<td>True</td>
<td>yes</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">9</td>
<td>ENSG00000137077.CCL21</td>
<td>precollector-LEC</td>
<td>pcLEC</td>
<td>True</td>
<td>yes</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">10</td>
<td>ENSG00000189056.RELN</td>
<td>precollector-LEC</td>
<td>pcLEC</td>
<td>True</td>
<td>yes</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">11</td>
<td>ENSG00000129048.ACKR4</td>
<td>precollector-LEC</td>
<td>pcLEC</td>
<td>False</td>
<td>yes</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">12</td>
<td>ENSG00000108691.CCL2</td>
<td>precollector-LEC</td>
<td>pcLEC</td>
<td>True</td>
<td>no</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">13</td>
<td>ENSG00000081041.CXCL2</td>
<td>precollector-LEC</td>
<td>pcLEC</td>
<td>True</td>
<td>no</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">14</td>
<td>ENSG00000133800.LYVE1</td>
<td>collector-LEC</td>
<td>clLEC</td>
<td>False</td>
<td>yes</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">15</td>
<td>ENSG00000137077.CCL21</td>
<td>collector-LEC</td>
<td>clLEC</td>
<td>False</td>
<td>yes</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">16</td>
<td>ENSG00000189056.RELN</td>
<td>collector-LEC</td>
<td>clLEC</td>
<td>False</td>
<td>yes</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">17</td>
<td>ENSG00000129048.ACKR4</td>
<td>collector-LEC</td>
<td>clLEC</td>
<td>True</td>
<td>yes</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">18</td>
<td>ENSG00000108691.CCL2</td>
<td>collector-LEC</td>
<td>clLEC</td>
<td>True</td>
<td>no</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">19</td>
<td>ENSG00000081041.CXCL2</td>
<td>collector-LEC</td>
<td>clLEC</td>
<td>True</td>
<td>no</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">20</td>
<td>ENSG00000137077.CCL21</td>
<td>valve-LEC</td>
<td>vLEC</td>
<td>False</td>
<td>yes</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">21</td>
<td>ENSG00000133800.LYVE1</td>
<td>valve-LEC</td>
<td>vLEC</td>
<td>False</td>
<td>yes</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">22</td>
<td>ENSG00000013297.CLDN11</td>
<td>valve-LEC</td>
<td>vLEC</td>
<td>True</td>
<td>yes</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">23</td>
<td>ENSG00000272398.CD24</td>
<td>valve-LEC</td>
<td>vLEC</td>
<td>True</td>
<td>no</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">24</td>
<td>ENSG00000149564.ESAM</td>
<td>valve-LEC</td>
<td>vLEC</td>
<td>True</td>
<td>no</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">25</td>
<td>ENSG00000170017.ALCAM</td>
<td>valve-LEC</td>
<td>vLEC</td>
<td>True</td>
<td>no</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">34</td>
<td>ENSG00000090339.ICAM1</td>
<td>LECs</td>
<td>LEC</td>
<td>True</td>
<td>?</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">35</td>
<td>ENSG00000103335.PIEZO1</td>
<td>LECs</td>
<td>LEC</td>
<td>True</td>
<td>?</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">36</td>
<td>ENSG00000154864.PIEZO2</td>
<td>LECs</td>
<td>LEC</td>
<td>True</td>
<td>?</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<div class="cell" data-execution_count="10">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># selecting genes highly variable in at least 2 batches and markers</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="co"># (markers are not HVG, but we need them)</span></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>selected_hvgs <span class="op">=</span> pearson_hvgs[pearson_hvgs.highly_variable_nbatches<span class="op">&gt;=</span><span class="dv">2</span>].index.union(markers.gene_ID)</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>selected_hvgs</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="10">
<pre><code>Index(['ENSG00000003436.TFPI', 'ENSG00000004799.PDK4', 'ENSG00000005483.KMT2E',
       'ENSG00000007908.SELE', 'ENSG00000008083.JARID2',
       'ENSG00000008294.SPAG9', 'ENSG00000008405.CRY1', 'ENSG00000008517.IL32',
       'ENSG00000008988.RPS20', 'ENSG00000009413.REV3L',
       ...
       'ENSG00000244694.PTCHD4', 'ENSG00000257987.TEX49',
       'ENSG00000265972.TXNIP', 'ENSG00000266964.FXYD1',
       'ENSG00000271503.CCL5', 'ENSG00000272398.CD24', 'ENSG00000276644.DACH1',
       'ENSG00000284491.THSD8', 'ENSG00000288602.C8orf44-SGK3',
       'ENSG00000291237.SOD2'],
      dtype='object', length=675)</code></pre>
</div>
</div>
<div class="cell" data-execution_count="11">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>adata.var[<span class="st">"selected"</span>] <span class="op">=</span> adata.var.index.isin(selected_hvgs)</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>adata.var</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="11">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">ID</th>
<th data-quarto-table-cell-role="th">Symbol</th>
<th data-quarto-table-cell-role="th">Type</th>
<th data-quarto-table-cell-role="th">n_cells_by_counts</th>
<th data-quarto-table-cell-role="th">mean_counts</th>
<th data-quarto-table-cell-role="th">log1p_mean_counts</th>
<th data-quarto-table-cell-role="th">pct_dropout_by_counts</th>
<th data-quarto-table-cell-role="th">total_counts</th>
<th data-quarto-table-cell-role="th">log1p_total_counts</th>
<th data-quarto-table-cell-role="th">selected</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">ENSG00000187634.SAMD11</td>
<td>ENSG00000187634</td>
<td>SAMD11</td>
<td>Gene Expression</td>
<td>21</td>
<td>0.000973</td>
<td>0.000972</td>
<td>99.902724</td>
<td>21.0</td>
<td>3.091042</td>
<td>False</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">ENSG00000188976.NOC2L</td>
<td>ENSG00000188976</td>
<td>NOC2L</td>
<td>Gene Expression</td>
<td>850</td>
<td>0.042709</td>
<td>0.041822</td>
<td>96.062627</td>
<td>922.0</td>
<td>6.827629</td>
<td>False</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">ENSG00000187961.KLHL17</td>
<td>ENSG00000187961</td>
<td>KLHL17</td>
<td>Gene Expression</td>
<td>108</td>
<td>0.005420</td>
<td>0.005405</td>
<td>99.499722</td>
<td>117.0</td>
<td>4.770685</td>
<td>False</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">ENSG00000187583.PLEKHN1</td>
<td>ENSG00000187583</td>
<td>PLEKHN1</td>
<td>Gene Expression</td>
<td>22</td>
<td>0.001019</td>
<td>0.001019</td>
<td>99.898092</td>
<td>22.0</td>
<td>3.135494</td>
<td>False</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">ENSG00000187642.PERM1</td>
<td>ENSG00000187642</td>
<td>PERM1</td>
<td>Gene Expression</td>
<td>3</td>
<td>0.000232</td>
<td>0.000232</td>
<td>99.986103</td>
<td>5.0</td>
<td>1.791759</td>
<td>False</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">ENSG00000271254.ENSG00000271254</td>
<td>ENSG00000271254</td>
<td>ENSG00000271254</td>
<td>Gene Expression</td>
<td>226</td>
<td>0.011349</td>
<td>0.011285</td>
<td>98.953122</td>
<td>245.0</td>
<td>5.505332</td>
<td>False</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">ENSG00000276345.ENSG00000276345</td>
<td>ENSG00000276345</td>
<td>ENSG00000276345</td>
<td>Gene Expression</td>
<td>491</td>
<td>0.025246</td>
<td>0.024932</td>
<td>97.725588</td>
<td>545.0</td>
<td>6.302619</td>
<td>False</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">ENSG00000276017.ENSG00000276017</td>
<td>ENSG00000276017</td>
<td>ENSG00000276017</td>
<td>Gene Expression</td>
<td>1</td>
<td>0.000046</td>
<td>0.000046</td>
<td>99.995368</td>
<td>1.0</td>
<td>0.693147</td>
<td>False</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">ENSG00000278817.ENSG00000278817</td>
<td>ENSG00000278817</td>
<td>ENSG00000278817</td>
<td>Gene Expression</td>
<td>189</td>
<td>0.008894</td>
<td>0.008855</td>
<td>99.124514</td>
<td>192.0</td>
<td>5.262690</td>
<td>False</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">ENSG00000277196.ENSG00000277196</td>
<td>ENSG00000277196</td>
<td>ENSG00000277196</td>
<td>Gene Expression</td>
<td>3</td>
<td>0.000139</td>
<td>0.000139</td>
<td>99.986103</td>
<td>3.0</td>
<td>1.386294</td>
<td>False</td>
</tr>
</tbody>
</table>

<p>18627 rows × 10 columns</p>
</div>
</div>
</div>
<div class="cell" data-execution_count="12">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># now working just on the HVGs</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>adata_hvgs <span class="op">=</span> adata[:,adata.var.selected].copy()</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>sc.pp.calculate_qc_metrics(adata_hvgs, inplace<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>adata_hvgs</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="12">
<pre><code>AnnData object with n_obs × n_vars = 21588 × 661
    obs: 'Sample', 'Barcode', 'donor', 'tissue', 'tissue_origin', 'sequencing_run', 'sex', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes'
    var: 'ID', 'Symbol', 'Type', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts', 'selected'
    uns: 'X_name'
    layers: 'logcounts', 'logcounts.scaled'</code></pre>
</div>
</div>
<div class="cell" data-execution_count="13">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># compute PCA, neighbors and UMAP just on HVGs</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>adata_hvgs.obsm[<span class="st">"logcounts.scaled.pca"</span>] <span class="op">=</span> sc.pp.pca(adata_hvgs.layers[<span class="st">"logcounts.scaled"</span>], svd_solver<span class="op">=</span><span class="st">"arpack"</span>, n_comps<span class="op">=</span><span class="dv">100</span>)</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>adata_hvgs</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="13">
<pre><code>AnnData object with n_obs × n_vars = 21588 × 661
    obs: 'Sample', 'Barcode', 'donor', 'tissue', 'tissue_origin', 'sequencing_run', 'sex', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes'
    var: 'ID', 'Symbol', 'Type', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts', 'selected'
    uns: 'X_name'
    obsm: 'logcounts.scaled.pca'
    layers: 'logcounts', 'logcounts.scaled'</code></pre>
</div>
</div>
<div class="cell" data-execution_count="14">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># 2 PCs</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>sc.pl.embedding(adata_hvgs, basis<span class="op">=</span><span class="st">"logcounts.scaled.pca"</span>, color<span class="op">=</span>[<span class="st">"donor"</span>, <span class="st">"sequencing_run"</span>, <span class="st">"tissue_origin"</span>, <span class="st">"tissue"</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>/anaconda3/envs/scampi/lib/python3.10/site-packages/scanpy/plotting/_tools/scatterplots.py:394: UserWarning: No data for colormapping provided via 'c'. Parameters 'cmap' will be ignored
  cax = scatter(
/anaconda3/envs/scampi/lib/python3.10/site-packages/scanpy/plotting/_tools/scatterplots.py:394: UserWarning: No data for colormapping provided via 'c'. Parameters 'cmap' will be ignored
  cax = scatter(
/anaconda3/envs/scampi/lib/python3.10/site-packages/scanpy/plotting/_tools/scatterplots.py:394: UserWarning: No data for colormapping provided via 'c'. Parameters 'cmap' will be ignored
  cax = scatter(
/anaconda3/envs/scampi/lib/python3.10/site-packages/scanpy/plotting/_tools/scatterplots.py:394: UserWarning: No data for colormapping provided via 'c'. Parameters 'cmap' will be ignored
  cax = scatter(</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="integration_n_clustering_files/figure-html/cell-15-output-2.png" class="img-fluid"></p>
</div>
</div>
<div class="cell" data-execution_count="15">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># harmony integration</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>sc.external.pp.harmony_integrate(</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>    adata_hvgs, basis<span class="op">=</span><span class="st">"logcounts.scaled.pca"</span>, key<span class="op">=</span><span class="st">"donor"</span>,</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>    adjusted_basis<span class="op">=</span><span class="st">"logcounts.scaled.pca.harmony"</span>, sigma<span class="op">=</span><span class="fl">1.</span> <span class="co">#hyperparameters: sigma.</span></span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a>adata_hvgs</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>2023-11-27 11:46:08,399 - harmonypy - INFO - Computing initial centroids with sklearn.KMeans...
2023-11-27 11:46:15,279 - harmonypy - INFO - sklearn.KMeans initialization complete.
2023-11-27 11:46:15,341 - harmonypy - INFO - Iteration 1 of 10
2023-11-27 11:46:17,578 - harmonypy - INFO - Iteration 2 of 10
2023-11-27 11:46:19,135 - harmonypy - INFO - Converged after 2 iterations</code></pre>
</div>
<div class="cell-output cell-output-display" data-execution_count="15">
<pre><code>AnnData object with n_obs × n_vars = 21588 × 661
    obs: 'Sample', 'Barcode', 'donor', 'tissue', 'tissue_origin', 'sequencing_run', 'sex', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes'
    var: 'ID', 'Symbol', 'Type', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts', 'selected'
    uns: 'X_name', 'donor_colors', 'sequencing_run_colors', 'tissue_origin_colors', 'tissue_colors'
    obsm: 'logcounts.scaled.pca', 'logcounts.scaled.pca.harmony'
    layers: 'logcounts', 'logcounts.scaled'</code></pre>
</div>
</div>
<div class="cell" data-execution_count="16">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>sc.pp.neighbors(adata_hvgs, use_rep<span class="op">=</span><span class="st">"logcounts.scaled.pca.harmony"</span>, n_pcs<span class="op">=</span><span class="dv">100</span>, n_neighbors<span class="op">=</span><span class="dv">40</span>, key_added<span class="op">=</span><span class="st">"logcounts.scaled.pca.harmony.neighbors"</span>)</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>adata_hvgs.obsm[<span class="st">"logcounts.scaled.pca.harmony.neighbors.umap"</span>] <span class="op">=</span> sc.tl.umap(adata_hvgs, neighbors_key<span class="op">=</span><span class="st">"logcounts.scaled.pca.harmony.neighbors"</span>, copy<span class="op">=</span><span class="va">True</span>).obsm[<span class="st">"X_umap"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>/anaconda3/envs/scampi/lib/python3.10/site-packages/umap/distances.py:1063: NumbaDeprecationWarning: The 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.
  @numba.jit()
/anaconda3/envs/scampi/lib/python3.10/site-packages/umap/distances.py:1071: NumbaDeprecationWarning: The 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.
  @numba.jit()
/anaconda3/envs/scampi/lib/python3.10/site-packages/umap/distances.py:1086: NumbaDeprecationWarning: The 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.
  @numba.jit()
/anaconda3/envs/scampi/lib/python3.10/site-packages/umap/umap_.py:660: NumbaDeprecationWarning: The 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.
  @numba.jit()</code></pre>
</div>
</div>
<div class="cell" data-execution_count="17">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>sc.pl.embedding(adata_hvgs, basis<span class="op">=</span><span class="st">"logcounts.scaled.pca.harmony"</span>, color<span class="op">=</span>[<span class="st">"donor"</span>, <span class="st">"sequencing_run"</span>, <span class="st">"tissue_origin"</span>, <span class="st">"tissue"</span>])</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>sc.pl.embedding(adata_hvgs, basis<span class="op">=</span><span class="st">"logcounts.scaled.pca.harmony.neighbors.umap"</span>, color<span class="op">=</span>[<span class="st">"donor"</span>, <span class="st">"sequencing_run"</span>, <span class="st">"tissue_origin"</span>, <span class="st">"tissue"</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>/anaconda3/envs/scampi/lib/python3.10/site-packages/scanpy/plotting/_tools/scatterplots.py:394: UserWarning: No data for colormapping provided via 'c'. Parameters 'cmap' will be ignored
  cax = scatter(
/anaconda3/envs/scampi/lib/python3.10/site-packages/scanpy/plotting/_tools/scatterplots.py:394: UserWarning: No data for colormapping provided via 'c'. Parameters 'cmap' will be ignored
  cax = scatter(
/anaconda3/envs/scampi/lib/python3.10/site-packages/scanpy/plotting/_tools/scatterplots.py:394: UserWarning: No data for colormapping provided via 'c'. Parameters 'cmap' will be ignored
  cax = scatter(
/anaconda3/envs/scampi/lib/python3.10/site-packages/scanpy/plotting/_tools/scatterplots.py:394: UserWarning: No data for colormapping provided via 'c'. Parameters 'cmap' will be ignored
  cax = scatter(
/anaconda3/envs/scampi/lib/python3.10/site-packages/scanpy/plotting/_tools/scatterplots.py:394: UserWarning: No data for colormapping provided via 'c'. Parameters 'cmap' will be ignored
  cax = scatter(
/anaconda3/envs/scampi/lib/python3.10/site-packages/scanpy/plotting/_tools/scatterplots.py:394: UserWarning: No data for colormapping provided via 'c'. Parameters 'cmap' will be ignored
  cax = scatter(
/anaconda3/envs/scampi/lib/python3.10/site-packages/scanpy/plotting/_tools/scatterplots.py:394: UserWarning: No data for colormapping provided via 'c'. Parameters 'cmap' will be ignored
  cax = scatter(
/anaconda3/envs/scampi/lib/python3.10/site-packages/scanpy/plotting/_tools/scatterplots.py:394: UserWarning: No data for colormapping provided via 'c'. Parameters 'cmap' will be ignored
  cax = scatter(</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="integration_n_clustering_files/figure-html/cell-18-output-2.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="integration_n_clustering_files/figure-html/cell-18-output-3.png" class="img-fluid"></p>
</div>
</div>
<div class="cell" data-execution_count="18">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co"># laiden clustering at different resolutions on the harmony integration layer</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>sc.tl.leiden(adata_hvgs, obsp<span class="op">=</span><span class="st">"logcounts.scaled.pca.harmony.neighbors_connectivities"</span>, resolution<span class="op">=</span><span class="fl">.5</span>, key_added<span class="op">=</span><span class="st">"logcounts.scaled.pca.harmony.neighbors_connectivities.leiden_05res"</span>)</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>sc.tl.leiden(adata_hvgs, obsp<span class="op">=</span><span class="st">"logcounts.scaled.pca.harmony.neighbors_connectivities"</span>, resolution<span class="op">=</span><span class="fl">.75</span>, key_added<span class="op">=</span><span class="st">"logcounts.scaled.pca.harmony.neighbors_connectivities.leiden_075res"</span>)</span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>adata_hvgs.obs</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="18">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">Sample</th>
<th data-quarto-table-cell-role="th">Barcode</th>
<th data-quarto-table-cell-role="th">donor</th>
<th data-quarto-table-cell-role="th">tissue</th>
<th data-quarto-table-cell-role="th">tissue_origin</th>
<th data-quarto-table-cell-role="th">sequencing_run</th>
<th data-quarto-table-cell-role="th">sex</th>
<th data-quarto-table-cell-role="th">n_genes_by_counts</th>
<th data-quarto-table-cell-role="th">log1p_n_genes_by_counts</th>
<th data-quarto-table-cell-role="th">total_counts</th>
<th data-quarto-table-cell-role="th">log1p_total_counts</th>
<th data-quarto-table-cell-role="th">pct_counts_in_top_50_genes</th>
<th data-quarto-table-cell-role="th">pct_counts_in_top_100_genes</th>
<th data-quarto-table-cell-role="th">pct_counts_in_top_200_genes</th>
<th data-quarto-table-cell-role="th">pct_counts_in_top_500_genes</th>
<th data-quarto-table-cell-role="th">logcounts.scaled.pca.harmony.neighbors_connectivities.leiden_05res</th>
<th data-quarto-table-cell-role="th">logcounts.scaled.pca.harmony.neighbors_connectivities.leiden_075res</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">donor1_fat.AAAGAACCACATACGT-1</td>
<td>donor1_fat</td>
<td>AAAGAACCACATACGT-1</td>
<td>1.0</td>
<td>fat</td>
<td>abdomen</td>
<td>1.0</td>
<td>F</td>
<td>252</td>
<td>5.533389</td>
<td>1317.0</td>
<td>7.183871</td>
<td>61.655277</td>
<td>81.624905</td>
<td>96.051632</td>
<td>100.0</td>
<td>0</td>
<td>1</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">donor1_fat.AAAGGTAGTCCGGTCA-1</td>
<td>donor1_fat</td>
<td>AAAGGTAGTCCGGTCA-1</td>
<td>1.0</td>
<td>fat</td>
<td>abdomen</td>
<td>1.0</td>
<td>F</td>
<td>139</td>
<td>4.941642</td>
<td>1479.0</td>
<td>7.299797</td>
<td>80.459770</td>
<td>97.363083</td>
<td>100.000000</td>
<td>100.0</td>
<td>0</td>
<td>2</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">donor1_fat.AAAGGTATCTCTCTTC-1</td>
<td>donor1_fat</td>
<td>AAAGGTATCTCTCTTC-1</td>
<td>1.0</td>
<td>fat</td>
<td>abdomen</td>
<td>1.0</td>
<td>F</td>
<td>196</td>
<td>5.283204</td>
<td>1806.0</td>
<td>7.499423</td>
<td>72.369878</td>
<td>91.251384</td>
<td>100.000000</td>
<td>100.0</td>
<td>1</td>
<td>3</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">donor1_fat.AACCACACAGTGGCTC-1</td>
<td>donor1_fat</td>
<td>AACCACACAGTGGCTC-1</td>
<td>1.0</td>
<td>fat</td>
<td>abdomen</td>
<td>1.0</td>
<td>F</td>
<td>254</td>
<td>5.541264</td>
<td>3110.0</td>
<td>8.042699</td>
<td>66.495177</td>
<td>88.938907</td>
<td>98.263666</td>
<td>100.0</td>
<td>0</td>
<td>1</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">donor1_fat.AACGGGATCTGCTGAA-1</td>
<td>donor1_fat</td>
<td>AACGGGATCTGCTGAA-1</td>
<td>1.0</td>
<td>fat</td>
<td>abdomen</td>
<td>1.0</td>
<td>F</td>
<td>171</td>
<td>5.147494</td>
<td>1448.0</td>
<td>7.278629</td>
<td>73.273481</td>
<td>93.784530</td>
<td>100.000000</td>
<td>100.0</td>
<td>0</td>
<td>2</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">donor5_recovery.TTTGGTTTCGACCAAT-1</td>
<td>donor5_recovery</td>
<td>TTTGGTTTCGACCAAT-1</td>
<td>5.0</td>
<td>mixed</td>
<td>abdomen</td>
<td>4.0</td>
<td>M</td>
<td>163</td>
<td>5.099866</td>
<td>618.0</td>
<td>6.428105</td>
<td>67.475728</td>
<td>89.320388</td>
<td>100.000000</td>
<td>100.0</td>
<td>1</td>
<td>3</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">donor5_recovery.TTTGTTGGTCCTACGG-1</td>
<td>donor5_recovery</td>
<td>TTTGTTGGTCCTACGG-1</td>
<td>5.0</td>
<td>mixed</td>
<td>abdomen</td>
<td>4.0</td>
<td>M</td>
<td>143</td>
<td>4.969813</td>
<td>305.0</td>
<td>5.723585</td>
<td>63.934426</td>
<td>85.901639</td>
<td>100.000000</td>
<td>100.0</td>
<td>0</td>
<td>2</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">donor5_recovery.TTTGTTGGTCGCAGTC-1</td>
<td>donor5_recovery</td>
<td>TTTGTTGGTCGCAGTC-1</td>
<td>5.0</td>
<td>mixed</td>
<td>abdomen</td>
<td>4.0</td>
<td>M</td>
<td>207</td>
<td>5.337538</td>
<td>471.0</td>
<td>6.156979</td>
<td>53.715499</td>
<td>76.433121</td>
<td>98.513800</td>
<td>100.0</td>
<td>5</td>
<td>7</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">donor5_recovery.TTTGTTGGTGGCGTAA-1</td>
<td>donor5_recovery</td>
<td>TTTGTTGGTGGCGTAA-1</td>
<td>5.0</td>
<td>mixed</td>
<td>abdomen</td>
<td>4.0</td>
<td>M</td>
<td>142</td>
<td>4.962845</td>
<td>813.0</td>
<td>6.701960</td>
<td>77.121771</td>
<td>94.833948</td>
<td>100.000000</td>
<td>100.0</td>
<td>3</td>
<td>8</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">donor5_recovery.TTTGTTGTCGGCATTA-1</td>
<td>donor5_recovery</td>
<td>TTTGTTGTCGGCATTA-1</td>
<td>5.0</td>
<td>mixed</td>
<td>abdomen</td>
<td>4.0</td>
<td>M</td>
<td>216</td>
<td>5.379897</td>
<td>763.0</td>
<td>6.638568</td>
<td>57.798165</td>
<td>80.471822</td>
<td>97.903014</td>
<td>100.0</td>
<td>3</td>
<td>6</td>
</tr>
</tbody>
</table>

<p>21588 rows × 17 columns</p>
</div>
</div>
</div>
<div class="cell" data-execution_count="19">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="co"># number of clusters for each resolution</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>adata_hvgs.obs.loc[:,adata_hvgs.obs.columns.<span class="bu">str</span>.match(<span class="st">"logcounts\.scaled\.pca\.harmony\.neighbors_connectivities\.leiden_[0-9]+res$"</span>)].nunique()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="19">
<pre><code>logcounts.scaled.pca.harmony.neighbors_connectivities.leiden_05res     11
logcounts.scaled.pca.harmony.neighbors_connectivities.leiden_075res    19
dtype: int64</code></pre>
</div>
</div>
<div class="cell" data-execution_count="20">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>sc.pl.embedding(adata_hvgs, basis<span class="op">=</span><span class="st">"logcounts.scaled.pca.harmony.neighbors.umap"</span>, color<span class="op">=</span><span class="st">"logcounts.scaled.pca.harmony.neighbors_connectivities.leiden_05res"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>/anaconda3/envs/scampi/lib/python3.10/site-packages/scanpy/plotting/_tools/scatterplots.py:394: UserWarning: No data for colormapping provided via 'c'. Parameters 'cmap' will be ignored
  cax = scatter(</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="integration_n_clustering_files/figure-html/cell-21-output-2.png" class="img-fluid"></p>
</div>
</div>
<div class="cell" data-execution_count="21">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>sc.pl.embedding(adata_hvgs, basis<span class="op">=</span><span class="st">"logcounts.scaled.pca.harmony.neighbors.umap"</span>, color<span class="op">=</span><span class="st">"logcounts.scaled.pca.harmony.neighbors_connectivities.leiden_075res"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>/anaconda3/envs/scampi/lib/python3.10/site-packages/scanpy/plotting/_tools/scatterplots.py:394: UserWarning: No data for colormapping provided via 'c'. Parameters 'cmap' will be ignored
  cax = scatter(</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="integration_n_clustering_files/figure-html/cell-22-output-2.png" class="img-fluid"></p>
</div>
</div>
<div class="cell" data-execution_count="22">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="co">#save the adata object</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>adata_hvgs.write_h5ad(<span class="st">"../output/1.1-sce-after-contamination-removal-integration-clustering.h5ad"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>



</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>